rfcs/026-work-enricher/README.md

```diff
@@ -0,0 +1,160 @@
+
+In either case, denormalisation of works within the enricher would consist of:
+
+1. Fetching the relations for a given work
+2. Emitting a new work containing the relevant fields embedded (`parts`, `partOf`, `precededBy`, `succeededBy`)
+3. Traversing the relations and emitting works for these in turn.
+
+Part `3` here is required as we have two way relations encoded for particular fields: for example, if in the tree below there is a node added with path `A/1/Z`, we not only need to send this work downstream but also send work `A/1` (containing updated `parts`), and works `A/1/X` and `A/1/Y` (containing updated `succeededBy`).
+
+```
+A
+├── 1
+│   ├── X
+│   └── Y
+│
+└── 2
+```
+
+Generally, for the insertion, update or removal of any given node, we need to update its siblings, its parent, and the whole subtree containing its descendants. The worst case for this process is an update of the root node, in which case the whole archive needs to be denormalised, with the largest archives containing in the order of 1000 works.
+
+## Commonalities with Other Pipeline Stages
+
+The matcher and merger stages have similar underlying functionality to the enricher. Abstractly these stages involve:
+
+1) Receiving a work as input and applying domain knowledge to build a graph of connections to other works
+2) Using the input work and its graph to emit a new work at the output, including any required data from the relations
+
+This points to a potential universal solution within the pipeline, such as using a common service, library or data store to assist with some of these tasks. Additionally, there are potential similarities with upcoming concepts work, where there has been some investigation into how concepts would fit within a graph model.
+
+There are a few issues to bear in mind when considering how a shared solution would work:
+
+* Whilst stage 1 (currently the matcher) could potentially also involve analysis of archive relations, it is important as stated above that the denormalisation in the enricher takes place on a merged works to ensure data consistency. This indicates the merger and enricher stages being separate and accessing works from two separate works stores (unmerged and merged).
+
+* The matcher currently uses a custom DynamoDB graph implementation which was built specifically for the purpose. This was never meant to be a general solution and a rewrite would likely need to be involved in the building of some common system.
+
+* The graph element of both the matcher and relations work are relatively similar, but the denormalisation stages (the enricher and the merger) have less parallels. For example, the input to the merger is currently a group of work IDs to be merged with the stage having no knowledge of the graph representation, unlike the enricher which is likely to receive single works and needs access to the graph.
```

Comment:
Merger input is a set of a set of IDs - iirc this is necessary for the _un_-merging of works. Good point though.

Reply:
To be honest I don't think the input format is the most important difference between the merger and the proposed enricher. We could modify the merger to receive single works and read the graph but the business logic of merging and the logic of enriching are completely different and probably wouldn't need to be base on field by field rules